java/Mekhmet-Efe_Adzhar/src/com/zoolatech/lecture6/tasks/_3/UserRepository.java

```diff
@@ -0,0 +1,31 @@
+package com.zoolatech.lecture6.tasks._3;
+
+import java.util.stream.Collectors;
+
+class UserRepository {
+    UserCache userCache;
+    UserTable userTable;
+
+    public UserRepository(UserCache userCache, UserTable userTable) {
+        this.userCache = userCache;
+        this.userTable = userTable;
+    }
+
+    public String findUserEmail(String id) throws UserMissingException {
+        if (userCache.findUser(id).isPresent()) {
+            return userCache.findUser(id).stream()
+                    .filter(user -> !user.getUserId().isEmpty())
+                    .map(User::getEmail)
+                    .collect(Collectors.toList())
+                    .toString();
+        } else if (userTable.findUser(id).isPresent()) {
+            return userTable.findUser(id).stream()
+                    .filter(user -> !user.getUserId().isEmpty())
+                    .map(User::getEmail)
+                    .collect(Collectors.toList())
+                    .toString();
+        } else {
+            throw new UserMissingException();
```

Comment:
it would be better to pass a message with a user id to the exception here and remove a default constructor from the exception class